src/Lean/Elab/PreDefinition/Basic.lean

```diff
@@ -117,10 +117,13 @@ private def addNonRecAux (preDef : PreDefinition) (compile : Bool) (all : List N
     withSaveInfoContext do  -- save new env
       addTermInfo' preDef.ref (← mkConstWithLevelParams preDef.declName) (isBinder := true)
     applyAttributesOf #[preDef] AttributeApplicationTime.afterTypeChecking
-    if preDef.modifiers.isNoncomputable then
+    let didCompile ←
+      if compile && shouldGenCodeFor preDef then
+        compileDecl decl
+      else
+        pure false
+    if preDef.modifiers.isNoncomputable || !didCompile then
       modifyEnv fun env => addNoncomputable env preDef.declName
```

Comment:
This means that all theorems will now be marked as noncomputable, whereas before they were not so marked. While it's unclear how much that matters, I think it would be good to not change things unnecessarily (and moreover, it would be great if we could have sufficiently thorough tests here that such changes are trivial to notice!)